avocado/plugins/multiplex.py

```diff
@@ -115,6 +120,40 @@ def run(self, args):
             sys.exit(exit_codes.AVOCADO_ALL_OK)
 
         variants = multiplexer.MuxTree(mux_tree)
+        if args.list:
+            prefix = '  '
+            for (index, tpl) in enumerate(variants):
+                level = 0
+                if args.list_variant_id:
+                    log.info("variant: !mux")
+                    level += 1
+
+                lname = ''
+                for x in tpl:
+                    path = x.path[len('/run'):]
+                    lname += path.replace('/', '-')
+                # trim first '-'
+                lname = lname[1:]
+                if args.list_variant_id:
+                    log.info('%s%d:' % (prefix * level, index))
+                    level += 1
+                log.info('%s%s:' % (prefix * level, lname))
+                level += 1
+                env = set()
+                for node in tpl:
+                    for key, value in node.environment.iteritems():
+                        origin = node.environment_origin[key].path
+                        env.add(("%s" % key, str(value)))
+                if not env:
+                    continue
+                fmt = '%s  %%-%ds: %%s' % (prefix * level,
+                                           max([len(_[0]) for _ in env]))
+                for record in sorted(env):
+                    log.info(fmt, *record)
+                log.info("")
+
+            sys.exit(exit_codes.AVOCADO_ALL_OK)
```

Comment:
In general I like this and we discussed something like this in relation to remote execution, which can utilize the same function to really dump the current Mux and send that file to the remote machine, which is better than sending the individual files.

Now about the results, it produces condensed yaml results, which I really dislike. What I mean is that this:

```yaml
hw:
    cpu: !mux
        intel:
            cpu_CFLAGS: '-march=core2'
        amd:
            cpu_CFLAGS: '-march=athlon64'
        arm:
            cpu_CFLAGS: '-mabi=apcs-gnu -march=armv8-a -mtune=arm8'
    disk: !mux
        scsi:
            disk_type: 'scsi'
        virtio:
            disk_type: 'virtio'
distro: !mux
    fedora:
        init: 'systemd'
    mint:
        init: 'systemv'
env: !mux
    debug:
        opt_CFLAGS: '-O0 -g'
    prod:
        opt_CFLAGS: '-O2'
```

becomes

```yaml
hw-cpu-intel-hw-disk-scsi-distro-fedora-env-debug:
    cpu_CFLAGS: -march=core2
    disk_type : scsi
    init      : systemd
    opt_CFLAGS: -O0 -g

hw-cpu-intel-hw-disk-scsi-distro-fedora-env-prod:
    cpu_CFLAGS: -march=core2
    disk_type : scsi
    init      : systemd
    opt_CFLAGS: -O2
...
```

and in multiplexer the node location (path inside params) is very important, but it's not kept. What I'd imagine as an output would be:

```
!mux
variant1:
    hw:
        cpu:
            intel:
                cpu_CFLAGS: '-march=core2'
        disk:
            scsi:
                disk_type: 'scsi'
    distro:
        fedora:
            init: 'systemd'
    env:
        debug:
            opt_CFLAGS: '-O0 -g'

variant2:
variant1:
    hw:
        cpu:
            amd:
                cpu_CFLAGS: '-march=athlon64'
        disk:
            scsi:
                disk_type: 'scsi'
    distro:
        fedora:
            init: 'systemd'
    env:
        debug:
            opt_CFLAGS: '-O0 -g'
```

Which would keep the paths. I'd even consider dumping it in a plain form:

```
---
# variant1
hw:
...
---
# variant2
hw:
---
```

Which would not work as a single yaml file, but it'd be simple to split it by `---` and use it in your tests. I'd probably support both, using those plain dumps and then `!mux`ed variants.

As for the variant names although I like the listing of all leaves, I'd consider using https://trello.com/c/1lzDCiqn/835-improve-the-variant-ids at least in the future. Btw good thing would be that in case of the single yaml files separated by `---` the variant-id hash would stay the same.

Last but not least I'd implement this inside the `avocado.core.tree` or `avocado.core.multiplexer` (not sure which one right now) and only used it from the `multiplex` plugin and later from the `remote` plugins.